src/tagmanager/tm_query.c

```diff
@@ -0,0 +1,314 @@
+ * You can only set this filter because the @a type is a bitmask containing
+ * multiple values.
+ *
+ * @param q The query to operate on.
+ * @param type Bitmasp of the tag time to filter, see @a TMTagType.
+ *
+ * @return 0 on succcess, < 0 on error.
+ */
+GEANY_API_SYMBOL
+gint tm_query_set_tag_type(TMQuery *q, TMTagType type)
+{
+	if (type > tm_tag_max_t)
+		return -1;
+
+	q->type = type;
+
+	return 0;
+}
+
+
+/* tm_tag_compare() expects pointer to pointers to tags, because it's
+ * used for GPtrArrays (therefore it's qsort-like). GQueue passes pointer
+ * to tags. */
+static gint tag_compare_ptr(gconstpointer a, gconstpointer b, gpointer data)
+{
+	return tm_tag_compare(&a, &b, data);
+}
+
+
+/** Execute a query
+ *
+ * The query will be executed according to the applied filters. Filters
+ * of the same type are combined using logical OR. Then, filters of different
+ * types are combined using logcal AND.
+ *
+ * The execution is stateless. Filters can be added after execution and the
+ * query can be re-run. Previous runs don't add side-effects.
+ *
+ * The results are generally sorted by tag name, even if @a sort_attr is NULL.
+ * supplied. The results can be
+ * sorted by @a sort_attr, and deduplicated by @a dedup_attr. Deduplication
+ * happens after sorting. There is an additional limitation for deduplication:
+ * Only adjacent tags can be deduplicated so they must be suitably ordered.
+ * This can be achieved by using the same attributes for @a sort_attr and
+ * @a dedup_attr.
+ *
+ * The results are stored in a @c GPtrArray. The individual tags are still
+ * owned by the workspace but the container array must be freed with
+ * g_ptr_array_free().
+ *
+ * @param q The query to execute
+ * @param sort_attr @nullable @array{zero-terminated=1} Tag attribute to use
+ *                   as sort key, or @c NULL to not sort.
+ * @param dedup_attr @nullable @array{zero-terminated=1} Tag attribute to use
+ *                    as deduplicaiton key, or @c NULL to not perform deduplicaiton.
+ * @return @transfer{container} @elementtype{TMTag} Resulting tag array.
+ */
+GEANY_API_SYMBOL
+GPtrArray *tm_query_exec(TMQuery *q, TMTagAttrType *sort_attr, TMTagAttrType *dedup_attr)
+{
+	TMTag **tags, *tag;
+	gint ntags;
+	guint i;
+	GString *s;
+	GQueue res = G_QUEUE_INIT;
+	TMSortOptions sort_options;
+	GList *node, *next;
+	gchar *scope;
+	TMParserType *lang;
+	GPtrArray *ret;
+
+	sort_options.sort_attrs = NULL;
+	/* tags_array isn not needed by tm_tag_compare(), but for tm_search_cmp() */
+	sort_options.tags_array = NULL;
+	sort_options.first = TRUE;
+
+	foreach_ptr_array(s, i, q->names)
+	{
+		TMTag **ptag;
+		sort_options.cmp_len = s->len;
+		if (q->data_sources & TM_QUERY_SOURCE_GLOBAL_TAGS)
+		{
+			tags = tm_tags_find(q->workspace->global_tags, s->str, s->len, &ntags);
+			foreach_c_array(ptag, tags, ntags)
+				g_queue_insert_sorted(&res, *ptag, tag_compare_ptr, &sort_options);
```

Comment:
> since the insertion point can be found with bisection

It couldn't as lists don't allow random access.  The only way is basically to traverse the whole list up to the insertion point.  If you knew the data set you could imagine making a guess at whether it's "close to" the end/start and start there, but that's as good as it gets AFAIK.

Sorting is a different story, because you can sort several items at once if you find interesting properties on them, and can take clever paths in some cases, but that doesn't work with a single insertion.

> The array still has the slight disadvantage that it must be resized when adding items exceeding the lenght, but this is not a problem here.

Yeah well, it's not so much worse than allocating the whole array at the end as `GPtrArray` is not so stupid and grows in chunks.